klickhouse/src/client.rs

```diff
@@ -289,6 +290,24 @@ impl Client {
         Self::connect_stream(read, writer, options).await
     }
 
+    /// Connects to a specific socket address over TLS (rustls) for Clickhouse.
+    /// This function will timeout after `duration` if the connection is not established.
+    #[cfg(feature = "tls")]
+    pub async fn connect_tls_timeout<A: ToSocketAddrs>(
+        destination: A,
+        options: ClientOptions,
+        name: rustls_pki_types::ServerName<'static>,
+        connector: &tokio_rustls::TlsConnector,
+        duration: Duration,
+    ) -> Result<Self> {
+        let destination = lookup_host(destination).await?.next().unwrap();
+        let tcp_stream = std::net::TcpStream::connect_timeout(&destination, duration)?;
```

Comment:
This is blocking behavior in an async context. Try using just the native Tokio TcpStream.

Reply:
Yes, but it it long-standing issue of Tokio's TcpStream that doesn't have feature to set socket connect timeout.
c.f. https://stackoverflow.com/questions/30022084/how-do-i-set-connect-timeout-on-tcpstream
In my understanding, there is no method provided to do both set connect_timeout AND async behaviour.
`let tcp_stream = async { std::net::TcpStream::connect_timeout(&destination, duration)?; }.await;` may be a solution?

Reply:
Yes, but it's a long-standing issue of Tokio's TcpStream, which does not have the ability to set a socket connect timeout. This PR implements the workaround described below.
https://stackoverflow.com/questions/30022084/how-do-i-set-connect-timeout-on-tcpstream
To my understanding, there is no method provided to do both set connect_timeout AND async behavior.
Perhaps `let tcp_stream = async { std::net::TcpStream::connect_timeout(&destination, duration)?; }.await;` might be a solution?

Reply:
Improved in aa35f0cd3c0ccf0a7006f2782422799517dab26f.